src/bugpoint.rs

```diff
@@ -601,52 +697,116 @@ fn reduce(
     mut func: Function,
     verbose: bool,
 ) -> Result<(Function, String), String> {
-    let mut ccc = CrashCheckContext::new(isa);
+    let mut context = CrashCheckContext::new(isa);
 
-    match ccc.check_for_crash(&func) {
+    match context.check_for_crash(&func) {
         CheckResult::Succeed => {
             return Err(format!(
-                "Given function compiled successfully or gave an verifier error."
+                "Given function compiled successfully or gave a verifier error."
             ));
         }
         CheckResult::Crash(_) => {}
     }
 
     resolve_aliases(&mut func);
 
+    let progress_bar = ProgressBar::with_draw_target(0, ProgressDrawTarget::stdout());
+    progress_bar.set_style(
+        ProgressStyle::default_bar().template("{bar:60} {prefix:40} {pos:>4}/{len:>4} {msg}"),
+    );
+
     for pass_idx in 0..100 {
         let mut should_keep_reducing = false;
         let mut phase = 0;
 
         loop {
-            let mut mutator = match phase {
-                0 => Box::new(RemoveInst::new(&func)) as Box<dyn Mutator>,
-                1 => Box::new(ReplaceInstWithIconst::new(&func)) as Box<dyn Mutator>,
-                2 => Box::new(ReplaceInstWithTrap::new(&func)) as Box<dyn Mutator>,
-                3 => Box::new(RemoveEbb::new(&func)) as Box<dyn Mutator>,
-                4 => Box::new(RemoveUnusedEntities::new()) as Box<dyn Mutator>,
+            let mut mutator: Box<dyn Mutator> = match phase {
+                0 => Box::new(RemoveInst::new(&func)),
+                1 => Box::new(ReplaceInstWithConst::new(&func)),
+                2 => Box::new(ReplaceInstWithTrap::new(&func)),
+                3 => Box::new(RemoveEbb::new(&func)),
+                4 => Box::new(RemoveUnusedEntities::new()),
+                5 => Box::new(MergeBlocks::new(&func)),
                 _ => break,
             };
 
-            func = mutator.reduce(
-                &mut ccc,
-                func,
-                format!("pass {}", pass_idx),
-                verbose,
-                &mut should_keep_reducing,
-            );
+            progress_bar.set_prefix(&format!("pass {} phase {}", pass_idx, mutator.name()));
+            progress_bar.set_length(mutator.mutation_count(&func) as u64);
+
+            // Reset progress bar.
+            progress_bar.set_position(0);
+            progress_bar.set_draw_delta(0);
+
+            for _ in 0..10000 {
+                progress_bar.inc(1);
+
+                let (mutated_func, msg, mutation_kind) = match mutator.mutate(func.clone()) {
+                    Some(res) => res,
+                    None => {
+                        break;
+                    }
+                };
+
+                if let ProgressStatus::Skip = mutation_kind {
+                    // The mutator didn't change anything, but we want to try more mutator
+                    // iterations.
+                    continue;
```

Comment:
Did you read my reaction at https://github.com/CraneStation/cranelift/commit/e110a3027d4cde53593129cedf938f9dcf8fcc17#r35503802?

Reply:
Yep. To be fair, this is a very minor point: in debug (non-optimized) builds, skipping states are non-visible, and in optimized + debug-assertions mode, everything is so fast that i couldn't see a single progress message.

At some point, I imagined that we could also have a way to disable progress messages at all, since they were likely to make reduction slower, but the optimized build convinced me it's not even useful to implement, since things are so fast right now.

Reply:
Are you fine with me merging like this? If you have a very strong opinion about this, I can move the code around.

Reply:
> Are you fine with me merging like this?

Yes

> At some point, I imagined that we could also have a way to disable progress messages at all

Makes sense, but I would like to retain the ability to show them. If it takes a lot of time to reduce, then it is useful to estimate the amount of time it will require. And it makes it possible to see when it is hanging. (for example infinite loop during legalization)